pkg/dogstatsd/server.go

```diff
@@ -318,22 +326,17 @@ func (s *Server) forwarder(fcon net.Conn, packetsChannel chan listeners.Packets)
 	}
 }
 
-func (s *Server) worker() {
-	// the batcher will be responsible of batching a few samples / events / service
-	// checks and it will automatically forward them to the aggregator, meaning that
-	// the flushing logic to the aggregator is actually in the batcher.
-	batcher := newBatcher(s.aggregator)
-
-	parser := newParser()
-	for {
-		select {
-		case <-s.stopChan:
-			return
-		case <-s.health.C:
-		case packets := <-s.packetsIn:
-			s.parsePackets(batcher, parser, packets)
-		}
-	}
+// Flush flushes all the data to the aggregator to them send it to the Datadog intake.
+// Set waitForSerializer to true to serialize and send the data synchronously.
+func (s *Server) Flush(waitForSerializer bool) {
+	log.Debug("Received a Flush trigger")
+	// make all workers flush their aggregated data (in the batcher) to the aggregator.
+	for _, w := range s.workers {
+		w.flush()
+	}
+	// flush the aggregator to have the serializer/forwarder send data to the backend.
+	// We add 10 seconds to the interval to ensure that we're getting the whole sketches bucket
+	s.aggregator.Flush(time.Now().Add(time.Second*10), true)
```

Comment:
2 comments:

1. the `waitForSerializer` argument isn't actually used, `true` is always passed to `aggregator.Flush`. I think we can remove `waitForSerializer` from `Server.Flush` entirely since I don't really see a use case to _not_ wait for the serializer here.
2. Adding 10 seconds to the time passed to `aggregator.Flush` seems a little hacky to me, because it relies implicitly on the internals of how sketch aggregation is implemented (and therefore it expects a behavior from the aggregator that's difficult to understand from the aggregator, and may have side effects on other parts of the aggregator). Instead, what do you think of passing a new argument instead that reflects whether the `timeSampler` should flush "buckets" that are incomplete? (for example named `flushIncompleteStatsdBuckets`). Also, flushing incomplete buckets only works well for sketch metrics, correct? (since the backend's sketch intake will be able to re-aggregate sketches on the same context and timestamp, whereas that won't work for the regular metrics intake). If I'm correct, then I think it makes sense to reflect in the variable name + behavior of the aggregator/timeSampler.